File: reference-implementations/validator-python/src/agent_protocol_validate/layer3.py
"""Layer 3 — drift detection. Rules 3.1 – 3.5.

3.2 consumes the per-bridge ``docs/bridges/<stack>-surface-map.yaml`` artifact
via :mod:`agent_protocol_validate.surface_map`. 3.4 is a plugin point with a
default cache-file implementation (``.agent-protocol/monitoring-cache.json``);
remote calls stay opt-in behind ``--enable-network``.
"""

from __future__ import annotations

import datetime as _dt
import json
import subprocess
from pathlib import Path
from typing import Any

from .findings import Finding
from .surface_map import SurfaceMap


def check(
    manifest: dict[str, Any],
    *,
    repo_root: Path,
    base_ref: str | None,
    surface_map: SurfaceMap | None = None,
    monitoring_cache: Path | None = None,
    uncontrolled_interface_max_age_days: int = 7,
    today: _dt.date | None = None,
) -> list[Finding]:
    if base_ref is None:
        return []
    today = today or _dt.date.today()

    changed = _git_diff_name_only(repo_root, base_ref)

    findings: list[Finding] = []
    findings.extend(_rule_3_1(manifest, changed))
    if surface_map is not None:
        findings.extend(_rule_3_2(manifest, surface_map, changed))
    findings.extend(_rule_3_3(manifest, surface_map, changed))
    findings.extend(_rule_3_4(manifest, monitoring_cache, uncontrolled_interface_max_age_days, today))
    findings.extend(_rule_3_5(manifest, repo_root, today))
    return findings


def _git_diff_name_only(repo_root: Path, base_ref: str) -> set[str]:
    try:
        out = subprocess.run(
            ["git", "diff", "--name-only", f"{base_ref}...HEAD"],
            cwd=repo_root,
            check=True,
            capture_output=True,
            text=True,
            timeout=30,
        )
    except (subprocess.CalledProcessError, FileNotFoundError, subprocess.TimeoutExpired):
        return set()
    return {line for line in out.stdout.splitlines() if line}


def _rule_3_1(manifest: dict[str, Any], changed: set[str]) -> list[Finding]:
    out: list[Finding] = []
    for sot in manifest.get("sot_map") or []:
        if not isinstance(sot, dict):
            continue
        if sot.get("role_in_change") == "read_only":
            continue
        src = sot.get("source")
        if not isinstance(src, str) or not src or src.startswith(("http://", "https://")):
            continue
        file_part = src.split(":", 1)[0]
        if file_part not in changed:
            out.append(
                Finding(
                    rule_id="drift.declared_sot_not_modified",
                    severity="advisory",
                    detail=f"declared SoT {file_part} not in diff",
                )
            )
    return out


def _rule_3_2(
    manifest: dict[str, Any], surface_map: SurfaceMap, changed: set[str]
) -> list[Finding]:
    out: list[Finding] = []
    for s in manifest.get("surfaces_touched") or []:
        if not isinstance(s, dict) or s.get("role") != "primary":
            continue
        name = s.get("surface")
        if not isinstance(name, str):
            continue
        patterns = surface_map.patterns_by_surface.get(name)
        if not patterns:
            continue
        if not any(name in surface_map.surfaces_for_path(f) for f in changed):
            out.append(
                Finding(
                    rule_id="drift.primary_surface_no_matching_file_change",
                    severity="advisory",
                    detail=(
                        f"surface {name!r} declared primary but no file matching "
                        f"{patterns} changed"
                    ),
                )
            )
    return out


def _rule_3_3(
    manifest: dict[str, Any], surface_map: SurfaceMap | None, changed: set[str]
) -> list[Finding]:
    cross = manifest.get("cross_cutting") or {}
    risk = cross.get("build_time_risk") or {}
    if not risk.get("codegen_touched"):
        return []
    if not risk.get("codegen_artifacts_committed"):
        return []
    if surface_map is None:
        return []
    patterns = surface_map.patterns_by_surface.get("__generated__") or []
    if not patterns:
        return []
    from .surface_map import _glob_match

    if any(any(_glob_match(f, p) for p in patterns) for f in changed):
        return []
    return [
        Finding(
            rule_id="drift.codegen_touched_but_no_generated_diff",
            severity="advisory",
            detail="codegen_touched=true but no generated artifacts in diff",
        )
    ]


def _rule_3_4(
    manifest: dict[str, Any],
    cache_path: Path | None,
    max_age_days: int,
    today: _dt.date,
) -> list[Finding]:
    if not cache_path or not cache_path.exists():
        return []
    try:
        cache = json.loads(cache_path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return []
    if not isinstance(cache, dict):
        return []
    out: list[Finding] = []
    for iface in manifest.get("uncontrolled_interfaces") or []:
        if not isinstance(iface, dict):
            continue
        channel = iface.get("monitoring_channel")
        if not isinstance(channel, str) or not channel:
            continue
        entry = cache.get(channel)
        if not isinstance(entry, dict):
            continue
        stamp = entry.get("last_check")
        try:
            last = _dt.date.fromisoformat(str(stamp)[:10])
        except ValueError:
            continue
        age_days = (today - last).days
        if age_days > max_age_days:
            out.append(
                Finding(
                    rule_id="drift.uncontrolled_interface_not_recently_checked",
                    severity="advisory",
                    detail=(
                        f"monitoring_channel={channel} last checked {age_days}d ago "
                        f"(> {max_age_days}d)"
                    ),
                )
            )
    return out


def _rule_3_5(
    manifest: dict[str, Any], repo_root: Path, today: _dt.date
) -> list[Finding]:
    last_updated = manifest.get("last_updated")
    if not last_updated:
        return []
    try:
        manifest_date = _dt.date.fromisoformat(str(last_updated)[:10])
    except ValueError:
        return []
    try:
        out = subprocess.run(
            ["git", "log", "-1", "--format=%cI"],
            cwd=repo_root,
            check=True,
            capture_output=True,
            text=True,
            timeout=10,
        )
    except (subprocess.CalledProcessError, FileNotFoundError, subprocess.TimeoutExpired):
        return []
    raw = out.stdout.strip()
    if not raw:
        return []
    try:
        latest_commit_date = _dt.date.fromisoformat(raw[:10])
    except ValueError:
        return []
    if manifest_date < latest_commit_date:
        return [
            Finding(
                rule_id="drift.manifest_older_than_latest_code",
                severity="advisory",
                detail=(
                    f"manifest.last_updated={manifest_date} older than latest code "
                    f"commit {latest_commit_date}"
                ),
            )
        ]
    return []

File: reference-implementations/community/validator-node/src/layer3.ts
// Layer 3 — drift detection. Rules 3.1 – 3.5.

import { spawnSync } from "node:child_process";
import { readFileSync, existsSync } from "node:fs";

import { makeFinding, type Finding } from "./findings.js";
import type { Manifest } from "./loader.js";
import { globMatch, type SurfaceMap } from "./surfaceMap.js";

export interface Layer3Options {
  repoRoot: string;
  baseRef?: string | null;
  surfaceMap?: SurfaceMap | null;
  monitoringCache?: string | null;
  uncontrolledInterfaceMaxAgeDays?: number;
  today?: Date;
}

export function check(manifest: Manifest, opts: Layer3Options): Finding[] {
  const baseRef = opts.baseRef ?? null;
  if (!baseRef) return [];
  const today = opts.today ?? new Date();
  const maxAge = opts.uncontrolledInterfaceMaxAgeDays ?? 7;

  const changed = gitDiffNameOnly(opts.repoRoot, baseRef);
  const out: Finding[] = [];
  out.push(...rule3_1(manifest, changed));
  if (opts.surfaceMap) {
    out.push(...rule3_2(manifest, opts.surfaceMap, changed));
  }
  out.push(...rule3_3(manifest, opts.surfaceMap ?? null, changed));
  out.push(...rule3_4(manifest, opts.monitoringCache ?? null, maxAge, today));
  out.push(...rule3_5(manifest, opts.repoRoot, today));
  return out;
}

function gitDiffNameOnly(repoRoot: string, baseRef: string): Set<string> {
  const res = spawnSync(
    "git",
    ["diff", "--name-only", `${baseRef}...HEAD`],
    { cwd: repoRoot, encoding: "utf-8", timeout: 30_000 }
  );
  if (res.status !== 0 || !res.stdout) return new Set();
  return new Set(res.stdout.split("\n").filter(Boolean));
}

function rule3_1(manifest: Manifest, changed: Set<string>): Finding[] {
  const out: Finding[] = [];
  const sots = (manifest["sot_map"] as unknown[]) || [];
  for (const s of sots) {
    if (!s || typeof s !== "object" || Array.isArray(s)) continue;
    const rec = s as Record<string, unknown>;
    if (rec["role_in_change"] === "read_only") continue;
    const src = rec["source"];
    if (
      typeof src !== "string" ||
      !src ||
      src.startsWith("http://") ||
      src.startsWith("https://")
    ) continue;
    const filePart = src.split(":", 1)[0]!;
    if (!changed.has(filePart)) {
      out.push(
        makeFinding(
          "drift.declared_sot_not_modified",
          "advisory",
          `declared SoT ${filePart} not in diff`
        )
      );
    }
  }
  return out;
}

export function rule3_2(
  manifest: Manifest,
  surfaceMap: SurfaceMap,
  changed: Set<string>
): Finding[] {
  const out: Finding[] = [];
  const surfaces = (manifest["surfaces_touched"] as unknown[]) || [];
  for (const s of surfaces) {
    if (!s || typeof s !== "object" || Array.isArray(s)) continue;
    const rec = s as Record<string, unknown>;
    if (rec["role"] !== "primary") continue;
    const name = rec["surface"];
    if (typeof name !== "string") continue;
    const patterns = surfaceMap.patternsBySurface[name];
    if (!patterns || patterns.length === 0) continue;
    const matched = [...changed].some(f => surfaceMap.surfacesForPath(f).includes(name));
    if (!matched) {
      out.push(
        makeFinding(
          "drift.primary_surface_no_matching_file_change",
          "advisory",
          `surface ${JSON.stringify(name)} declared primary but no file matching ${JSON.stringify(patterns)} changed`
        )
      );
    }
  }
  return out;
}

function rule3_3(
  manifest: Manifest,
  surfaceMap: SurfaceMap | null,
  changed: Set<string>
): Finding[] {
  const cross = (manifest["cross_cutting"] as Record<string, unknown>) || {};
  const risk = (cross["build_time_risk"] as Record<string, unknown>) || {};
  if (!risk["codegen_touched"]) return [];
  if (!risk["codegen_artifacts_committed"]) return [];
  if (!surfaceMap) return [];
  const patterns = surfaceMap.patternsBySurface["__generated__"];
  if (!patterns || patterns.length === 0) return [];
  const hit = [...changed].some(f => patterns.some(p => globMatch(f, p)));
  if (hit) return [];
  return [
    makeFinding(
      "drift.codegen_touched_but_no_generated_diff",
      "advisory",
      "codegen_touched=true but no generated artifacts in diff"
    ),
  ];
}

export function rule3_4(
  manifest: Manifest,
  cachePath: string | null,
  maxAgeDays: number,
  today: Date
): Finding[] {
  if (!cachePath || !existsSync(cachePath)) return [];
  let cache: unknown;
  try {
    cache = JSON.parse(readFileSync(cachePath, "utf-8"));
  } catch {
    return [];
  }
  if (!cache || typeof cache !== "object" || Array.isArray(cache)) return [];
  const out: Finding[] = [];
  const todayMs = Date.UTC(today.getUTCFullYear(), today.getUTCMonth(), today.getUTCDate());
  const ifaces = (manifest["uncontrolled_interfaces"] as unknown[]) || [];
  for (const i of ifaces) {
    if (!i || typeof i !== "object" || Array.isArray(i)) continue;
    const channel = (i as Record<string, unknown>)["monitoring_channel"];
    if (typeof channel !== "string" || !channel) continue;
    const entry = (cache as Record<string, unknown>)[channel];
    if (!entry || typeof entry !== "object") continue;
    const stamp = (entry as Record<string, unknown>)["last_check"];
    if (typeof stamp !== "string") continue;
    const m = stamp.match(/^(\d{4})-(\d{2})-(\d{2})/);
    if (!m) continue;
    const lastMs = Date.UTC(Number(m[1]), Number(m[2]) - 1, Number(m[3]));
    const ageDays = Math.floor((todayMs - lastMs) / 86_400_000);
    if (ageDays > maxAgeDays) {
      out.push(
        makeFinding(
          "drift.uncontrolled_interface_not_recently_checked",
          "advisory",
          `monitoring_channel=${channel} last checked ${ageDays}d ago (> ${maxAgeDays}d)`
        )
      );
    }
  }
  return out;
}

function rule3_5(manifest: Manifest, repoRoot: string, today: Date): Finding[] {
  const lastUpdated = manifest["last_updated"];
  if (!lastUpdated) return [];
  const md = String(lastUpdated).match(/^(\d{4})-(\d{2})-(\d{2})/);
  if (!md) return [];
  const manifestMs = Date.UTC(Number(md[1]), Number(md[2]) - 1, Number(md[3]));
  const res = spawnSync(
    "git",
    ["log", "-1", "--format=%cI"],
    { cwd: repoRoot, encoding: "utf-8", timeout: 10_000 }
  );
  if (res.status !== 0) return [];
  const raw = res.stdout.trim();
  const cd = raw.match(/^(\d{4})-(\d{2})-(\d{2})/);
  if (!cd) return [];
  const commitMs = Date.UTC(Number(cd[1]), Number(cd[2]) - 1, Number(cd[3]));
  if (manifestMs < commitMs) {
    const fmt = (ms: number) => {
      const d = new Date(ms);
      const y = d.getUTCFullYear();
      const mo = String(d.getUTCMonth() + 1).padStart(2, "0");
      const da = String(d.getUTCDate()).padStart(2, "0");
      return `${y}-${mo}-${da}`;
    };
    return [
      makeFinding(
        "drift.manifest_older_than_latest_code",
        "advisory",
        `manifest.last_updated=${fmt(manifestMs)} older than latest code commit ${fmt(commitMs)}`
      ),
    ];
  }
  // keep `today` referenced to silence unused-param lint
  void today;
  return [];
}
